File: src/mvmctl/core/console/_controller.py
"""Console controller for VM serial console access.

Provides high-level console relay management.
"""

import os
from pathlib import Path

from mvmctl.exceptions import ConsoleError
from mvmctl.services.console_relay import ConsoleRelayClient, ConsoleRelayManager


class ConsoleController:
    """High-level console relay interface for a specific VM.

    Combines relay manager (lifecycle) and client (connection) for
    easy console operations. Manages PTY lifecycle internally.
    """

    def __init__(
        self,
        vm_id: str,
        vm_dir: Path,
        vm_name: str | None = None,
        pid_filename: str = "console.pid",
        socket_filename: str = "console.sock",
        log_filename: str = "firecracker.console.log",
    ) -> None:
        """Initialize console relay for a VM.

        Args:
            vm_id: VM unique identifier
            vm_dir: VM directory path
            vm_name: Human-readable name (uses vm_id if None)
            pid_filename: Name of the PID file
            socket_filename: Name of the socket file
            log_filename: Name of the log file
        """
        self._manager = ConsoleRelayManager(
            id=vm_id,
            path=vm_dir,
            name=vm_name or vm_id,
            pid_filename=pid_filename,
            socket_filename=socket_filename,
            log_filename=log_filename,
        )
        self._client: ConsoleRelayClient | None = None
        self._controller_fd: int | None = None
        self._client_fd: int | None = None
        self._pid: int | None = None
        self._socket_path: Path | None = None

    @property
    def controller_fd(self) -> int | None:
        """Access controller FD (for relay)."""
        return self._controller_fd

    @property
    def client_fd(self) -> int | None:
        """Access client FD (for Firecracker)."""
        return self._client_fd

    @property
    def manager(self) -> ConsoleRelayManager:
        """Access the underlying relay manager."""
        return self._manager

    @property
    def socket_path(self) -> Path | None:
        return self._socket_path

    @property
    def pid(self) -> int | None:
        return self._pid

    def create_pty(self) -> int:
        """Lazy PTY creation - returns client FD for Firecracker.

        Creates PTY only when first called. Subsequent calls return
        the already-created client FD.

        Returns:
            Client file descriptor for Firecracker stdin/stdout

        Raises:
            OSError: If PTY creation fails
            ConsoleError: If PTY client FD is None after creation
        """
        if self._controller_fd is None:
            self._controller_fd, self._client_fd = os.openpty()

        if self._client_fd is None:
            raise ConsoleError("PTY allocation failed: client FD is None after creation")

        return self._client_fd

    def close_client_fd(self) -> None:
        """Close client FD after Firecracker has taken ownership.

        Safe to call multiple times. Called by orchestrator after
        Firecracker process is spawned.
        """
        if self._client_fd is not None:
            try:
                os.close(self._client_fd)
            except OSError:
                pass
            self._client_fd = None

    def close_pty(self) -> None:
        """Close both PTY FDs. For cleanup on error paths."""
        self.close_client_fd()

        if self._controller_fd is not None:
            try:
                os.close(self._controller_fd)
            except OSError:
                pass
            self._controller_fd = None

    def start(self) -> tuple[Path, int]:
        """Start the console relay for this VM.

        Must call create_pty() before this method.

        Returns:
            Tuple of (socket_path, pid)

        Raises:
            RuntimeError: If create_pty() was not called first
        """
        if self._controller_fd is None:
            raise RuntimeError("Must call create_pty() before start()")

        self._socket_path, self._pid = self._manager.start(self._controller_fd)

        return self._socket_path, self._pid

    def cleanup(self) -> None:
        self.stop()
        self.close_pty()
        self.close_client_fd()

    def stop(self) -> None:
        """Stop the console relay gracefully."""
        self._manager.stop()

    def terminate(self) -> bool:
        """Forcefully terminate the relay.

        Returns:
            True if terminated, False if not running
        """
        return self._manager.terminate()

    def is_running(self) -> bool:
        """Check if relay is running."""
        return self._manager.is_running()

    def get_pid(self) -> int | None:
        """Get relay PID if running."""
        return self._manager.get_pid()

    def connect(self) -> ConsoleRelayClient:
        """Connect to the console.

        Returns:
            Connected ConsoleRelayClient
        """
        self._client = ConsoleRelayClient(self._manager.socket_path)
        self._client.connect()
        return self._client

    def disconnect(self) -> None:
        """Disconnect from console."""
        if self._client:
            self._client.disconnect()
            self._client = None

File: internal/core/console/controller.go
// Package console provides VM serial console output streaming.
// Layer: Core domain — never imports other core/* packages.
package console

import (
	"context"
	"errors"
	"fmt"
	"os"
	"path/filepath"

	"golang.org/x/sys/unix"

	consolesvc "mvmctl/internal/service/console"
)

// Controller manages the console lifecycle for a single VM.
//
// The Controller owns:
// - PTY pair creation/teardown (master → relay, slave → Firecracker)
// - Relay subprocess spawning/management (Start, Stop, IsRunning, GetPID)
// - Relay client connection (Connect, Disconnect)
//
// The API layer creates a Controller and calls these methods — it never
// touches the service package directly.
//
// A Controller is not safe for concurrent use. The API layer serializes
// calls per-VM through its own execution model.
type Controller struct {
	vmID, vmPath, vmName        string
	pidFilename, socketFilename string

	// PTY state
	masterFD int // PTY master → relay subprocess
	clientFD int // PTY slave → Firecracker
	hasPTY   bool

	// Relay (set by Start, used by Stop/IsRunning/PID/SocketPath/Connect)
	relayManager *consolesvc.Relay

	// Client (set by Connect)
	client *consolesvc.RelayClient
}

// NewController creates a new Controller for the given VM.
func NewController(vmID, vmPath, vmName, pidFilename, socketFilename string) *Controller {
	if vmName == "" {
		vmName = vmID
	}
	return &Controller{
		vmID:           vmID,
		vmPath:         vmPath,
		vmName:         vmName,
		pidFilename:    pidFilename,
		socketFilename: socketFilename,
	}
}

// CreatePTY creates a PTY pair and returns the slave/client FD.
func (cc *Controller) CreatePTY() (int, error) {
	if cc.hasPTY {
		return cc.clientFD, nil
	}

	ptmx, err := os.OpenFile("/dev/ptmx", os.O_RDWR, 0)
	if err != nil {
		return 0, fmt.Errorf("failed to open /dev/ptmx: %w", err)
	}

	masterFD := int(ptmx.Fd())
	ptyno, err := unix.IoctlGetInt(masterFD, unix.TIOCGPTN)
	if err != nil {
		ptmx.Close()
		return 0, fmt.Errorf("failed to get PTY slave number: %w", err)
	}
	slaveName := fmt.Sprintf("/dev/pts/%d", ptyno)

	if err := unix.IoctlSetPointerInt(masterFD, unix.TIOCSPTLCK, 0); err != nil {
		ptmx.Close()
		return 0, fmt.Errorf("failed to unlock PTY: %w", err)
	}

	slave, err := os.OpenFile(slaveName, os.O_RDWR, 0)
	if err != nil {
		ptmx.Close()
		return 0, fmt.Errorf("failed to open slave PTY %s: %w", slaveName, err)
	}

	cc.masterFD = masterFD
	cc.clientFD = int(slave.Fd())
	cc.hasPTY = true

	if cc.clientFD == 0 {
		ptmx.Close()
		slave.Close()
		cc.masterFD = 0
		cc.hasPTY = false
		return 0, errors.New("PTY allocation failed: client FD is None after creation")
	}

	return cc.clientFD, nil
}

// Start starts the console relay for this controller's VM.
func (cc *Controller) Start(ctx context.Context) (string, *int, error) {
	if !cc.hasPTY {
		return "", nil, errors.New("Must call create_pty() before start()")
	}

	ptyFile := os.NewFile(uintptr(cc.masterFD), "pty")
	if ptyFile == nil {
		return "", nil, fmt.Errorf("invalid PTY controller FD %d", cc.masterFD)
	}

	cfg := consolesvc.Config{
		VMID:           cc.vmID,
		VMPath:         cc.vmPath,
		VMName:         cc.vmName,
		PIDFilename:    cc.pidFilename,
		SocketFilename: cc.socketFilename,
	}

	result, err := consolesvc.Spawn(ctx, cfg, ptyFile)
	if err != nil {
		return "", nil, err
	}

	// Spawn closed ptyFile (parent's copy of the PTY master fd). Clear
	// masterFD to prevent ClosePTY/Cleanup from double-closing.
	cc.masterFD = 0

	// Create the relay manager now that the subprocess is alive.
	// Spawn already wrote the PID file, so the Relay can find the PID.
	pidPath := filepath.Join(cc.vmPath, cc.pidFilename)
	sockPath := filepath.Join(cc.vmPath, cc.socketFilename)
	cc.relayManager = consolesvc.NewRelay(cc.vmName, pidPath, sockPath)

	return result.SocketPath, &result.PID, nil
}

// SocketPath returns the relay socket path (set after Start).
func (cc *Controller) SocketPath() string {
	if cc.relayManager == nil {
		return ""
	}
	return cc.relayManager.SocketPath()
}

// Connect connects to the console relay and returns a RelayClient.
func (cc *Controller) Connect() (*consolesvc.RelayClient, error) {
	if cc.client != nil && cc.client.IsConnected() {
		return cc.client, nil
	}

	if cc.relayManager == nil {
		return nil, errors.New("relay not started: call Start() before Connect()")
	}

	socketPath := cc.relayManager.SocketPath()
	client := consolesvc.NewRelayClient(socketPath, nil)
	if err := client.Connect(); err != nil {
		return nil, err
	}

	cc.client = client
	return client, nil
}

// Disconnect disconnects from the console relay.
func (cc *Controller) Disconnect() error {
	if cc.client != nil {
		cc.client.Disconnect()
		cc.client = nil
	}
	return nil
}

// Stop stops the console relay.
func (cc *Controller) Stop(force bool) bool {
	if cc.relayManager == nil {
		return false
	}
	return cc.relayManager.Stop(force)
}

// IsRunning returns true if the relay is currently running.
func (cc *Controller) IsRunning() bool {
	if cc.relayManager == nil {
		return false
	}
	return cc.relayManager.IsRunning()
}

// GetPID returns the relay PID and whether it's running.
func (cc *Controller) GetPID() (int, bool) {
	if cc.relayManager == nil {
		return 0, false
	}
	return cc.relayManager.PIDAlive()
}

// CloseClientFD closes the PTY slave/client file descriptor.
func (cc *Controller) CloseClientFD() {
	if cc.clientFD != 0 {
		_ = unix.Close(cc.clientFD)
		cc.clientFD = 0
	}
}

// ClosePTY closes both PTY file descriptors.
func (cc *Controller) ClosePTY() {
	if cc.clientFD != 0 {
		_ = unix.Close(cc.clientFD)
		cc.clientFD = 0
	}
	if cc.masterFD != 0 {
		_ = unix.Close(cc.masterFD)
		cc.masterFD = 0
	}
	cc.hasPTY = false
}

// Cleanup stops the relay gracefully and closes PTY FDs.
func (cc *Controller) Cleanup() {
	if cc.relayManager != nil {
		cc.relayManager.Stop(false)
	}

	if cc.clientFD != 0 {
		_ = unix.Close(cc.clientFD)
		cc.clientFD = 0
	}
	if cc.masterFD != 0 {
		_ = unix.Close(cc.masterFD)
		cc.masterFD = 0
	}
	cc.hasPTY = false
}
